Support the TIME type in the Iceberg connector

djsstarburst opened the pull request:
This commit adds support for the TIME type in the Iceberg
connector.  The TIME type is stored as microsecond time since
midnight UTC in Parquet and ORC, so it occupies an int64.

The main change was adding Parquet and ORC column readers and writers
for TIME that convert between Presto engine picosecond time
and Iceberg connector native microsecond time.

martint commented:
We should wait until this goes in https://github.com/prestosql/presto/pull/4381

The time type is currently very broken. That PR adds support for variable precision times and fixes the semantics and representation of the type

djsstarburst commented:
> We should wait until this goes in #4381

Ok, we'll be waiting :)

djsstarburst commented:
Since #4381 is now merged, I rebased this PR on tip master, and updated it to be consistent with picosecond TIME in the Presto engine.  This PR is now ready for review.

electrum commented on presto-parquet/src/main/java/io/prestosql/parquet/writer/valuewriter/TimeValueWriter.java:
```diff
@@ -0,0 +1,50 @@
+    {
+        super(parquetType, valuesWriter);
+        this.type = requireNonNull(type, "type is null");
+        this.scaleValue = parquetType.isPrimitive() && parquetType.getOriginalType() == OriginalType.TIME_MICROS;
```
Name the class `TimeMicrosValueWriter` and check the logical type outside, then we don't need to specialize inside the writer code.

djsstarburst replied:
Done.  It's great to remove that grotty check in the writer!

electrum commented on presto-parquet/src/main/java/io/prestosql/parquet/writer/valuewriter/TimeValueWriter.java:
```diff
@@ -0,0 +1,50 @@
+        for (int i = 0; i < block.getPositionCount(); i++) {
+            if (!block.isNull(i)) {
+                long value = type.getLong(block, i);
+                long scaledValue = scaleValue ? value / Timestamps.PICOSECONDS_PER_MICROSECOND : value;
```
`roundDiv`

djsstarburst replied:
Per @martint's comment, left as is.

electrum commented on presto-orc/src/main/java/io/prestosql/orc/writer/TimeColumnWriter.java:
```diff
@@ -0,0 +1,43 @@
+    @Override
+    protected long transformValue(long value)
+    {
+        return value / PICOSECONDS_PER_MICROSECOND;
```
`roundDiv`

djsstarburst replied:
Per @martint's comment, left as is.

electrum commented on presto-orc/src/main/java/io/prestosql/orc/reader/TimeColumnReader.java:
```diff
@@ -0,0 +1,39 @@
+    protected void maybeTransformValues(long[] values, int nextBatchSize)
+    {
+        for (int i = 0; i < nextBatchSize; i++) {
+            values[i] = values[i] * PICOSECONDS_PER_MICROSECOND;
```
Can use `*=`

djsstarburst replied:
Changed.

electrum commented on presto-orc/src/main/java/io/prestosql/orc/writer/ColumnWriters.java:
```diff
@@ -47,6 +48,9 @@ public static ColumnWriter createColumnWriter(
     {
         requireNonNull(type, "type is null");
+        if (type instanceof TimeType) {
+            return new TimeColumnWriter(columnId, type, compression, bufferSize, () -> new IntegerStatisticsBuilder(bloomFilterBuilder.get()));
```
```java
checkArgument(orcType == LONG, "wrong ORC type for time type: %s", orcType);

djsstarburst replied:
Added.

martint commented on presto-iceberg/src/main/java/io/prestosql/plugin/iceberg/ExpressionConverter.java:
```diff
@@ -181,6 +183,10 @@ private static Object getIcebergLiteralValue(Type type, Marker marker)
         }
 
+        if (type instanceof TimeType) {
+            return ((long) marker.getValue()) / PICOSECONDS_PER_MICROSECOND;
```
TIME type doesn't use epoch (
it's just picoseconds since midnight), and its values cannot be negative, so this is unnecessary.

djsstarburst replied:
Left as it is.

electrum replied:
We still need rounding, right?

djsstarburst replied:
@electrum, I should assume the question is directed at @martint, right?

djsstarburst commented on presto-iceberg/src/main/java/io/prestosql/plugin/iceberg/IcebergPageSink.java:
```diff
@@ -361,6 +363,9 @@ public static Object getIcebergValue(Block block, int position, Type type)
             return type.getDouble(block, position);
         }
+        if (type instanceof TimeType) {
+            return type.getLong(block, position) / PICOSECONDS_PER_MICROSECOND;
```
Based on @martint's previous comment, left as it is.

djsstarburst commented on presto-iceberg/src/main/java/io/prestosql/plugin/iceberg/IcebergSplitSource.java:
```diff
@@ -126,12 +124,18 @@ private ConnectorSplit toIcebergSplit(FileScanTask task)
-                }
-                else {
-                    partitionValue = value.toString();
+                switch (type.typeId()) {
```
Done, and much nicer.  Since in `String.valueOf((Long) value)`, the cast to `Long` is redundant, the former TIME case is correctly handled by the default case, as pointed out by Intellij.

djsstarburst commented on presto-iceberg/src/main/java/io/prestosql/plugin/iceberg/PartitionTable.java:
```diff
@@ -300,6 +301,9 @@ private static Object convert(Object value, Type type)
             return utcMillis;
         }
+        if (type instanceof Types.TimeType) {
+            return ((Long) value) * Timestamps.PICOSECONDS_PER_MICROSECOND;
```
Done.

electrum commented on presto-orc/src/main/java/io/prestosql/orc/writer/ColumnWriters.java:
```diff
@@ -48,6 +51,10 @@ public static ColumnWriter createColumnWriter(
     {
         requireNonNull(type, "type is null");
         OrcType orcType = orcTypes.get(columnId);
+        if (type instanceof TimeType) {
```
Add a comment that this is for Iceberg

djsstarburst replied:
Added.

electrum commented on presto-orc/src/main/java/io/prestosql/orc/TupleDomainOrcPredicate.java:
```diff
@@ -189,6 +191,9 @@ public static Domain getDomain(Type type, long rowCount, ColumnStatistics column
 
         boolean hasNullValue = columnStatistics.getNumberOfValues() != rowCount;
 
+        if (type instanceof TimeType && columnStatistics.getIntegerStatistics() != null) {
```
Comment that this is for Iceberg

djsstarburst replied:
Added.

electrum commented on presto-iceberg/src/main/java/io/prestosql/plugin/iceberg/IcebergSplitSource.java:
```diff
@@ -125,18 +123,22 @@ private ConnectorSplit toIcebergSplit(FileScanTask task)
+
+    private static String partitionValueFromType(Type type, Object value)
+    {
+        switch (type.typeId()) {
```
I think you accidentally removed the case for `TIME`

djsstarburst replied:
> Since in String.valueOf((Long) value), the cast to Long is redundant, the former TIME case is correctly handled by the default case, as pointed out by Intellij.

No; as mentioned in this PR comment above, the default case suffices

electrum commented on presto-iceberg/src/main/java/io/prestosql/plugin/iceberg/TypeConverter.java:
```diff
@@ -176,6 +177,9 @@ public static Type toPrestoType(org.apache.iceberg.types.Type type, TypeManager
         if (type instanceof MapType) {
             return fromMap((MapType) type);
         }
+        if (type instanceof TimeType) {
```
This should only allow `time(6)`. We should probably add a `TIME_MICROS` constant for `createTimeType(6)` like I did in the timestamps PR. Then add another check
```java
if (type instanceof TimestampType) {
    throw new PrestoException(NOT_SUPPORTED, format("Time precision (%s) not supported for Iceberg. Use \"time(6)\" instead.", ((TimeType) type).getPrecision()));
}

djsstarburst replied:
Done.

electrum commented on presto-orc/src/main/java/io/prestosql/orc/reader/ColumnReaders.java:
```diff
@@ -32,6 +33,9 @@ public static ColumnReader createColumnReader(
             OrcBlockFactory blockFactory)
             throws OrcCorruptionException
     {
+        if (type instanceof TimeType) {
```
I realized why this is problematic. In the ORC code, which is generic, we making a huge assumption that the `time` type is written as a `long` with microsecond encoding. We should be able to do this in a safe way by checking the `iceberg.long-type = TIME` attribute from `OrcColumn`. The code would still have this ugly Iceberg specific logic, but it would be safe if we introduce other usages of `TimeType` in the future.

djsstarburst replied:
I made this change and force-pushed.

electrum replied:
Verify three things:
* `precision == 6`
* `column.getColumnType() == LONG`
* `"LONG".equals(column.getAttributes("iceberg.long-type"))`

electrum commented on presto-iceberg/src/main/java/io/prestosql/plugin/iceberg/PartitionTransforms.java:
```diff
@@ -253,6 +257,27 @@ private static Block extractTimestamp(Block block, LongUnaryOperator function)
+        return extractHoursFromTime(block, HOURS_DURATION::getValueAsLong);
+    }
+
+    private static Block extractHoursFromTime(Block block, LongUnaryOperator function)
```
Name this `extractTime` since it's not specific to hours

djsstarburst replied:
Changed.

electrum commented on presto-iceberg/src/main/java/io/prestosql/plugin/iceberg/IcebergSplitSource.java:
```diff
@@ -125,18 +123,22 @@ private ConnectorSplit toIcebergSplit(FileScanTask task)
         return Collections.unmodifiableMap(partitionKeys);
     }
+
+    private static String partitionValueFromType(Type type, Object value)
```
It looks like this didn't require any changes, so we can leave it as-is

djsstarburst replied:
Change reverted.

electrum commented on presto-iceberg/src/main/java/io/prestosql/plugin/iceberg/ExpressionConverter.java:
```diff
@@ -181,6 +183,10 @@ private static Object getIcebergLiteralValue(Type type, Marker marker)
             return toIntExact(((Long) marker.getValue()));
         }
 
+        if (type instanceof TimeType) {
```
`TIME_MICROS`

djsstarburst replied:
Changed.

electrum commented on presto-iceberg/src/main/java/io/prestosql/plugin/iceberg/PartitionData.java:
```diff
@@ -134,6 +134,8 @@ public static Object getValue(JsonNode partitionValue, Type type)
             case LONG:
             case TIMESTAMP:
                 return partitionValue.asLong();
+            case TIME:
```
Can collapse to previous cases

djsstarburst replied:
Collapsed.

electrum commented on presto-iceberg/src/main/java/io/prestosql/plugin/iceberg/TypeConverter.java:
```diff
@@ -261,6 +270,9 @@ private static TypeInfo toHiveTypeInfo(Type type)
         if (DATE.equals(type)) {
             return HIVE_DATE.getTypeInfo();
         }
+        if (type.equals(TIME_MICROS)) {
```
Flip `equals()` so it's consistent with the others

djsstarburst replied:
Flipped.

electrum commented on presto-iceberg/src/test/java/io/prestosql/plugin/iceberg/TestIcebergSmoke.java:
```diff
@@ -144,6 +145,35 @@ private void testDecimalWithPrecisionAndScale(Session session, FileFormat format
+    {
+        String tableName = format("test_%s%s_time", partitioned ? "partitioned_" : "", format.name().toLowerCase(ENGLISH));
+        String partitioning = partitioned ? ", partitioning = ARRAY['x']" : "";
+        assertUpdate(session, format("CREATE TABLE %s (x TIME, y BIGINT) WITH (format = '%s'%s)", tableName, format.name().toLowerCase(ENGLISH), partitioning));
```
No need to lowercase format

djsstarburst replied:
Removed.

electrum commented on presto-orc/src/main/java/io/prestosql/orc/writer/TimeColumnWriter.java:
```diff
@@ -0,0 +1,43 @@
+public class TimeColumnWriter
+        extends LongColumnWriter
+{
+    public TimeColumnWriter(
```
No need to wrap here, especially for simple delegation

djsstarburst replied:
Unwrapped.

electrum commented on presto-parquet/src/main/java/io/prestosql/parquet/writer/valuewriter/TimeMicrosValueWriter.java:
```diff
@@ -0,0 +1,47 @@
+        for (int i = 0; i < block.getPositionCount(); i++) {
+            if (!block.isNull(i)) {
+                long value = type.getLong(block, i);
+                long scaledValue = value / Timestamps.PICOSECONDS_PER_MICROSECOND;
```
Static import

We could inline this above
```java
long value = type.getLong(block, i) / PICOSECONDS_PER_MICROSECOND;

djsstarburst replied:
Inlined.

djsstarburst commented:
This TIME PR is rebased on master so it now has no merge conflicts, @electrum

electrum commented:
Thanks!
